// leaf-utils/functions.js
const { MessageActionRow, MessageButton } = require('discord.js');

const converter = ['ᴀ', 'ʙ', 'ᴄ', 'ᴅ', 'ᴇ', 'ꜰ', 'ɢ', 'ʜ', 'ɪ', 'ᴊ',
  'ᴋ', 'ʟ', 'ᴍ', 'ɴ', 'ᴏ', 'ᴘ', 'ǫ', 'ʀ', 'ꜱ', 'ᴛ', 'ᴜ', 'ᴠ', 'ᴡ', 'x', 'ʏ', 'ᴢ'];
const letters = 'abcdefghijklmnopqrstuvwxyz'.split('');

module.exports = {
  i(length) {
    var randomChars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    var result = '';
    for (var i = 0; i < length; i++) {
      result += randomChars.charAt(Math.floor(Math.random() * randomChars.length))
    }
    return result;
  },

  shuffle(array) {
    for (var i = array.length - 1; i > 0; i--) {
      var j = Math.floor(Math.random() * (i + 1));
      var temp = array[i];
      array[i] = array[j];
      array[j] = temp;
    }
    return array;
  },

  smallCaps(text) {
    const content = text.toLowerCase().split('').map(letter => {
      if (/[a-z]/g.test(letter)) {
        var index = letters.indexOf(letter)
        return converter[index];
      } else {
        return letter
      }
    }).join('');
    return content;
  },

  reverseText(content) {
    return content.split('').reverse().join('');
  },

  getRandomString: function(length) {
		const randomChars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
		let result = '';
		for (let i = 0; i < length; i++) {
			result += randomChars.charAt(
				Math.floor(Math.random() * randomChars.length),
			);
		}
		return result;
	},

  convertTime: function(time) {
		const absoluteSeconds = Math.floor((time / 1000) % 60);
		const absoluteMinutes = Math.floor((time / (1000 * 60)) % 60);
		const absoluteHours = Math.floor((time / (1000 * 60 * 60)) % 24);
		const absoluteDays = Math.floor(time / (1000 * 60 * 60 * 24));
		const d = absoluteDays
			? absoluteDays === 1
				? '1 day'
				: `${absoluteDays} days`
			: null;
		const h = absoluteHours
			? absoluteHours === 1
				? '1 hour'
				: `${absoluteHours} hours`
			: null;
		const m = absoluteMinutes
			? absoluteMinutes === 1
				? '1 minute'
				: `${absoluteMinutes} minutes`
			: null;
		const s = absoluteSeconds
			? absoluteSeconds === 1
				? '1 second'
				: `${absoluteSeconds} seconds`
			: null;
		const absoluteTime = [];
		if (d) absoluteTime.push(d);
		if (h) absoluteTime.push(h);
		if (m) absoluteTime.push(m);
		if (s) absoluteTime.push(s);
		return absoluteTime.join(', ');
	},

  shuffleArray: function(array) {
		for (let i = array.length - 1; i > 0; i--) {
			const j = Math.floor(Math.random() * (i + 1));
			const temp = array[i];
			array[i] = array[j];
			array[j] = temp;
		}
		return array;
	},

  randomHexColor: function() {
		return (
			'#' +
			('000000' + Math.floor(Math.random() * 16777215).toString(16)).slice(-6)
		);
	},

  addRow: function(btns) {
		const row = new MessageActionRow();
		for (const btn of btns) {
			row.addComponents(btn);
		}
		return row;
	},

  createButton: function(label, disabled, getRandomString) {
    let style = 'SECONDARY';
    if (label === 'AC' || label === 'DC' || label === '⌫') {
      style = 'DANGER';
    } else if (label === '=') {
      style = 'SUCCESS';
    } else if (
      label === '(' ||
      label === ')' ||
      label === '^' ||
      label === '%' ||
      label === '÷' ||
      label === 'x' ||
      label === '-' ||
      label === '+' ||
      label === '.'
    ) {
      style = 'PRIMARY';
    }
    if (disabled) {
      const btn = new MessageButton()
        .setLabel(label)
        .setStyle(style)
        .setDisabled();
      if (label === '\u200b') {
        btn.setCustomId(getRandomString(10));
      } else {
        btn.setCustomId('cal' + label);
      }
      return btn;
    } else {
      const btn = new MessageButton().setLabel(label).setStyle(style);
      if (label === '\u200b') {
        btn.setDisabled();
        btn.setCustomId(getRandomString(10));
      } else {
        btn.setCustomId('cal' + label);
      }
      return btn;
    }
  }
}
