// aoi.mongo/dist/query/Length.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Length = void 0;
const Match_1 = require("./Match");
/** A filter which work with data length */
class Length {
    constructor(operator, path, filter, expr) {
        this.operator = operator;
        this.path = path;
        this.filter = filter;
        this.expr = expr;
    }
    /** Filter if length is equal */
    equal(length) {
        return this.compare("$eq", length);
    }
    /** Filter if length is not equal */
    notEqual(length) {
        return this.compare("$ne", length);
    }
    /** Filter if length is greater than */
    greaterThan(length) {
        return this.compare("$gt", length);
    }
    /** Filter if length is greater than or equal */
    greaterThanEqual(length) {
        return this.compare("$gte", length);
    }
    /** Filter if length is less than */
    lessThan(length) {
        return this.compare("$lt", length);
    }
    /** Filter if length is less than or equal */
    lessThanEqual(length) {
        return this.compare("$lte", length);
    }
    /** Filter if the comparison return true */
    compare(operator, length) {
        const match = new Match_1.Match(this.filter);
        this.expr[operator] = [
            {
                [this.operator]: this.path
            },
            {
                $literal: length
            }
        ];
        return match;
    }
}
exports.Length = Length;
