// aoi.mongo/dist/query/Number.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NumFilter = void 0;
const Match_1 = require("./Match");
/** A filter which work with numbers */
class NumFilter {
    constructor(path, filter, expr) {
        this.path = path;
        this.filter = filter;
        this.expr = expr;
    }
    /** Filter if number is equal */
    equal(num) {
        return this.compare("$eq", num);
    }
    /** Filter if number is not equal */
    notEqual(num) {
        return this.compare("$ne", num);
    }
    /** Filter if number is greater than */
    greaterThan(num) {
        return this.compare("$gt", num);
    }
    /** Filter if number is greater than or equal */
    greaterThanEqual(num) {
        return this.compare("$gte", num);
    }
    /** Filter if number is less than */
    lessThan(num) {
        return this.compare("$lt", num);
    }
    /** Filter if number is less than or equal */
    lessThanEqual(num) {
        return this.compare("$lte", num);
    }
    /** Filter if the comparison return true */
    compare(operator, num) {
        const match = new Match_1.Match(this.filter);
        this.expr[operator] = [this.path, {
                $literal: num
            }];
        return match;
    }
}
exports.NumFilter = NumFilter;
