// aoi.mongo/dist/query/Query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Query = void 0;
const Cursor_1 = require("../Cursor");
const Generator_1 = require("../Generator");
const Transformer_1 = require("../Transformer");
/** A wrapper class for querying */
class Query {
    constructor(mongo, match) {
        this.mongo = mongo;
        this.filter = match.filter;
    }
    /** Find one matched document */
    async findOne(options) {
        const doc = await this.mongo.collection.findOne(this.filter, options);
        return doc ? (0, Transformer_1.transform)(doc) : null;
    }
    /** Find all matched document */
    async findMulti(options) {
        const cursor = await this.mongo.collection.find(this.filter, options);
        return new Cursor_1.Cursor(cursor);
    }
    /** Update one matched document */
    updateOne(value, options) {
        const refs = [];
        const data = (0, Generator_1.generate)(value, refs);
        const filter = { $set: { data, refs } };
        if (options)
            options.upsert = false;
        if (options)
            return this.mongo.collection.updateOne(this.filter, filter, options);
        return this.mongo.collection.updateOne(this.filter, filter);
    }
    /** Update all matched document */
    updateMulti(value, options) {
        const refs = [];
        const data = (0, Generator_1.generate)(value, refs);
        const filter = { $set: { data, refs } };
        if (options)
            options.upsert = false;
        if (options)
            return this.mongo.collection.updateMany(this.filter, filter, options);
        return this.mongo.collection.updateMany(this.filter, filter);
    }
    /** Rename a matched document to another key */
    rename(key, options) {
        const filter = { $set: { key } };
        if (options)
            options.upsert = false;
        if (options)
            return this.mongo.collection.updateOne(this.filter, filter, options);
        return this.mongo.collection.updateOne(this.filter, filter);
    }
    /** Delete one matched document */
    deleteOne(options) {
        if (options)
            return this.mongo.collection.deleteOne(this.filter, options);
        return this.mongo.collection.deleteOne(this.filter);
    }
    /** Delete all matched document */
    deleteMulti(options) {
        if (options)
            return this.mongo.collection.deleteMany(this.filter, options);
        return this.mongo.collection.deleteMany(this.filter);
    }
}
exports.Query = Query;
